File: package/main/src/Random/weightedChoice.ts
export interface WeightedItem<T> {
  value: T;
  weight: number;
}

/**
 * Returns a random element using cumulative weights and binary search.
 * Items with non-positive weights are skipped from the pool.
 *
 * @template T - Element type
 * @param {readonly WeightedItem<T>[]} items - Items with non-negative weights
 * @returns {T} A randomly selected element
 * @example
 * weightedChoice([{ value: "a", weight: 1 }, { value: "b", weight: 4 }]);
 */
export const weightedChoice = <T>(items: readonly WeightedItem<T>[]): T => {
  const cumulative: number[] = [];
  let total = 0;
  for (const item of items) {
    if (item.weight > 0) {
      total += item.weight;
    }
    cumulative.push(total);
  }
  const target = Math.random() * total;
  let lo = 0;
  let hi = cumulative.length - 1;
  while (lo < hi) {
    const mid = (lo + hi) >>> 1;
    if (cumulative[mid] <= target) {
      lo = mid + 1;
    } else {
      hi = mid;
    }
  }
  return items[lo].value;
};

File: package/umt_python/src/random/weighted_choice.py
import random as random_module
from collections.abc import Sequence
from typing import Any, TypedDict


class WeightedItem(TypedDict):
    value: Any
    weight: float


def weighted_choice(items: Sequence[WeightedItem]) -> Any:  # noqa: ANN401
    """
    Returns a random element using cumulative weights and binary search.

    Items with non-positive weights are skipped from the pool.

    Args:
        items: Items with non-negative weights, each a mapping with
            ``value`` and ``weight`` keys.

    Returns:
        A randomly selected value.

    Example:
        >>> weighted_choice([{"value": "skip", "weight": 0}, {"value": "win", "weight": 5}])
        'win'
        >>> weighted_choice([{"value": "a", "weight": 1}, {"value": "b", "weight": 1}]) in ("a", "b")
        True
    """
    cumulative: list[float] = []
    total = 0.0
    for item in items:
        if item["weight"] > 0:
            total += item["weight"]
        cumulative.append(total)
    target = random_module.random() * total
    lo = 0
    hi = len(cumulative) - 1
    while lo < hi:
        mid = (lo + hi) >> 1
        if cumulative[mid] <= target:
            lo = mid + 1
        else:
            hi = mid
    return items[lo]["value"]
